src/cliff_vesting/src/main.nr

```diff
@@ -0,0 +1,141 @@
+                sender,
+                receiver, 
+                escrow,
+                vesting,
+                randomness,
+                header: NoteHeader::empty() 
+            }
+        }
+    }
+
+    pub fn filter_note_escrow(
+        notes: [Option<CliffVestingNote>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL],
+        escrow: AztecAddress,
+    ) -> [Option<CliffVestingNote>; MAX_NOTE_HASH_READ_REQUESTS_PER_CALL] {
+        let mut selected = [Option::none(); MAX_NOTE_HASH_READ_REQUESTS_PER_CALL];
+
+        for i in 0..notes.len() {
+            if notes[i].is_some() {
+                let note = notes[i].unwrap_unchecked();
+                if note.escrow == escrow {
+                    selected[0] = Option::some(note);
+                    selected
+                }
+            }
+        }
+
+        assert(false, "No cliff vesting found for escrow");
+    }
+
+    #[private]
+    fn create_cliff_vesting(escrow: AztecAddress, receiver: AztecAddress, vesting: U128) {
+        let escrow_owner = Escrow::at(escrow).owner();
+        assert(escrow_owner == context.this_address(), "Cliff vesting is not owner");
+
+        let sender = context.msg_sender();
+        let note = CliffVestingNote::new(
+            sender, 
+            receiver, 
+            escrow, 
+            vesting, 
+            context.randomness()
+        );
+
+        // Do we need to check if there is already a cliff vesting created for this escrow?
+        let note_insertion = storage.escrows.insert(&mut context, note);
+
+        // Emit the note for the sender
+        note_insertion.emit(encode_and_encrypt_note(
+            &mut context,
+            sender,
+            sender,
+        ));
+
+        // Emit the note for the receiver
+        note_insertion.emit(encode_and_encrypt_note(
+            &mut context,
+            receiver,
+            sender,
+        ));
+    }
+
+    #[private]
+    fn claim(escrow: AztecAddress) {
+        let caller = context.msg_sender();
+        
+        // Find and remove the note
+        // There should only be one note per escrow
+        let note = find_and_remove_cliff_note(escrow);
+            
+        assert(note.escrow == escrow, "Invalid escrow address");
+        assert(caller == note.receiver, "Only receiver can claim");
+
+        Escrow::at(escrow).withdraw(note.receiver).call(&mut context);
+    }
+
+    #[private]
+    fn clawback(escrow: AztecAddress) {
+        let caller = context.msg_sender();
+        
+        // Find and remove the note
+        // There should only be one note per escrow
+        let note = find_and_remove_cliff_note(escrow);
+            
+        assert(note.escrow == escrow, "Invalid escrow address");
+        assert(caller == note.sender, "Only sender can clawback");
+
+        Escrow::at(escrow).withdraw(caller).call(&mut context);
+    }
+
+    #[private]
+    fn find_and_remove_cliff_note(escrow: AztecAddress) -> CliffVestingNote {
```

Comment:
i think for "internal" calls we need to use `#[contract_library_method]`, because we're not expecting this method to be externally called (private entrypoint)